src/tpu.rs

```diff
@@ -33,14 +71,109 @@ impl Tpu {
         bank: &Arc<Bank>,
         tick_duration: Config,
         transactions_sockets: Vec<UdpSocket>,
+        broadcast_socket: UdpSocket,
+        cluster_info: Arc<RwLock<ClusterInfo>>,
+        entry_height: u64,
         sigverify_disabled: bool,
         max_tick_height: Option<u64>,
         last_entry_id: &Hash,
         leader_id: Pubkey,
-    ) -> (Self, Receiver<Vec<Entry>>, Arc<AtomicBool>) {
+        is_leader: bool,
+        to_validator_tx: TpuRotationSender,
+    ) -> Self {
         let exit = Arc::new(AtomicBool::new(false));
 
-        let (fetch_stage, packet_receiver) = FetchStage::new(transactions_sockets, exit.clone());
+        let tpu_mode = if is_leader {
+            let (fetch_stage, packet_receiver) =
+                FetchStage::new(transactions_sockets, exit.clone());
+
+            let (sigverify_stage, verified_receiver) =
+                SigVerifyStage::new(packet_receiver, sigverify_disabled);
+
+            let (banking_stage, entry_receiver) = BankingStage::new(
+                &bank,
+                verified_receiver,
+                tick_duration,
+                last_entry_id,
+                max_tick_height,
+                leader_id,
+                to_validator_tx,
+            );
+
+            let broadcast_service = BroadcastService::new(
+                bank.clone(),
+                broadcast_socket,
+                cluster_info.clone(),
+                entry_height,
+                bank.leader_scheduler.clone(),
+                entry_receiver,
+                max_tick_height,
+                exit.clone(),
+            );
+
+            let svcs = LeaderServices::new(
+                fetch_stage,
+                sigverify_stage,
+                banking_stage,
+                broadcast_service,
+            );
+            TpuMode::Leader(svcs)
+        } else {
+            let tpu_forwarder = TpuForwarder::new(transactions_sockets, cluster_info.clone());
+            let svcs = ForwarderServices::new(tpu_forwarder);
+            TpuMode::Forwarder(svcs)
+        };
+
+        let tpu = Self {
+            tpu_mode,
+            exit: exit.clone(),
+        };
+
+        tpu
+    }
+
+    pub fn switch_to_forwarder(
+        &mut self,
+        transactions_sockets: Vec<UdpSocket>,
+        cluster_info: Arc<RwLock<ClusterInfo>>,
+    ) {
+        match &self.tpu_mode {
+            TpuMode::Leader(svcs) => {
+                svcs.fetch_stage.close();
+            }
+            TpuMode::Forwarder(svcs) => {
```

Comment:
In what conditions switch_to_forwarder() be called on a node that's already a forwarder? If it's already a forwarder, should we leave it as is (instead of closing the forwarder and creating a new one)?

I have same question on switch_to_leader() function.

Reply:
For switch to leader, the tpu might have died when it reached the top(poh thread returns) of its slot but if there are no other validators or if the same node was selected to be the leader again, we need to restart the tpu.

I just left the behavior consistent for the forwarder although there should be no scenario where the forwarded needs a full restart like this.